File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include<bits/stdc++.h>
using namespace std;
bool convert(string &a, string &b)
{
    int count=0;
    for(int i=0;i<a.size();i++)
    {
        if(a[i]!=b[i]) count++;
        if(count>1) return false;
    }
    if(count==1) return true;
}
int solution(string begin, string target, vector<string> words) {
    if(find(words.begin(),words.end(),target)==words.end()) return 0; // words 단어 중 target이 없음
    queue<pair<string,int>> q; // 현재 단어, cnt
    vector<bool> visited(words.size(),0);
    
    q.push({begin,0});
    
    while(q.size()) // 탐색
    {
        string cur = q.front().first;
        int cnt = q.front().second;
        q.pop();
        
        if(cur==target) return cnt; 
        
        for(int i=0; i<words.size(); i++) // words 단어들과 비교
        {
            if(!visited[i] && convert(cur,words[i])) // 확인하지 않았고 변환이 가능하다면
            {
                visited[i]=true; // 방문 처리
                q.push({words[i],cnt+1}); // 갱신
            }
        }
    }
    return 0;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
def solution(begin, target, words):
    if target not in words:
        return 0

    n = len(words)
    visited = [False] * n
    ret = [10**9]  # INT_MAX 대체

    def convert(a, b):
        count = 0
        for i in range(len(a)):
            if a[i] != b[i]:
                count += 1
                if count > 1:
                    return False
        return True  # (<=1 글자 차이면 True) C++ 로직 그대로

    def dfs(cur, cnt):
        if cur == target:
            if cnt < ret[0]:
                ret[0] = cnt
            return
        for i in range(n):
            if not visited[i] and convert(cur, words[i]):
                visited[i] = True
                dfs(words[i], cnt + 1)
                visited[i] = False

    dfs(begin, 0)
    return 0 if ret[0] == 10**9 else ret[0]
